### ISSUE TITLE
ubuntu boxes can fail to upgrade sudo packages noninteratively

### ISSUE THREAD

**Issue by smammy** [chef/bento#711]
For Ubuntu boxes, Bento modifies the distribution-provided `/etc/sudoders` file to add `Defaults exempt_group=sudo` and to add `NOPASSWD` to the entry for `%sudo`. This is done in `scripts/ubuntu/sudoers.sh`.

My provisioning process runs `apt-get -qy update && apt-get -qy dist-upgrade` to get latest Ubuntu updates. On my current `bento/ubuntu-14.04-i386` box (v2.3.0), this upgrades the sudo package from 1.8.9p5-1ubuntu1.2 to 1.8.9p5-1ubuntu1.3, which includes a new version of `/etc/sudoers`. The upgrade fails because there is a conffile conflict to resove and the provisioner is non-interactive. Here is the relevant part of the provisioner output:

```
==> default: Setting up sudo (1.8.9p5-1ubuntu1.3) ...
==> default: Configuration file '/etc/sudoers'
==> default:  ==> Modified (by you or by a script) since installation.
==> default:  ==> Package distributor has shipped an updated version.
==> default:    What would you like to do about it ?  Your options are:
==> default:     Y or I  : install the package maintainer's version
==> default:     N or O  : keep your currently-installed version
==> default:       D     : show the differences between the versions
==> default:       Z     : start a shell to examine the situation
==> default:  The default action is to keep your current version.
==> default: *** sudoers (Y/I/N/O/D/Z) [default=N] ?
==> default: Configuration file '/etc/sudoers'
==> default:  ==> Modified (by you or by a script) since installation.
==> default:  ==> Package distributor has shipped an updated version.
==> default:    What would you like to do about it ?  Your options are:
==> default:     Y or I  : install the package maintainer's version
==> default:     N or O  : keep your currently-installed version
==> default:       D     : show the differences between the versions
==> default:       Z     : start a shell to examine the situation
==> default:  The default action is to keep your current version.
==> default: *** sudoers (Y/I/N/O/D/Z) [default=N] ?
==> default: dpkg: error processing package sudo (--configure):
==> default:  EOF on stdin at conffile prompt
```

I can't think of a graceful way to handle this on my end. Forcing 'Y' to this question would clobber bento's changes (which are helpful) and forcing 'N' could break things. (Not in this case, but you get the idea.)

Could `sudoers.sh` drop the two lines y'all need into e.g. `/etc/sudoers.d/vagrant` instead of editing `/etc/sudoers` itself?

**Comment by cheeseplus** [comment]
That is something we could certainly look at - typically you can pass apt options that like `--force-conf-old` or the like to preserve a certain behavior and still keep it automated.

**Comment by smammy** [comment]
Thanks for considering it. `--force-confold` won't quite work for me because I want to accept distributions updates to `/etc/sudoers`, just in case they include a security improvement or similar. My understanding is that Debian added the `/etc/sudoers.d` apparatus (in 1.7.2p1-1, incidentally) to solve just this kind of problem.

**Comment by w0rd-driven** [comment]
Just to give my findings: if you use `--force-confnew`, things like vagrant halt stops working because `NOPASSWD` is no longer there. I'm also only using `apt-get -y upgrade`, it doesn't have to be a `dist-upgrade` to be an issue.

I have no problem using `--force-confold` in this instance but not using `/etc/sudoers.d/vagrant` will become a problem for any downstream box that happens to use this as their base.

**Comment by cheeseplus** [comment]
This should be fairly straightforward and definitely a good things to minimize any unexpected behaviour.

**Comment by shiva** [comment]
Is this going to be included in some release? Also, I didn't quite understand, why sudo group should exempt from secure path? My understanding is the purpose of the secure_path, is to allow sudo users to execute binary in the /sbin, /usr/sbin etc, without forcing the normal users to be exposed to these paths/binaries.

**Comment by smammy** [comment]
For those following along at home, this was fixed in master by #832 on 30 Jun 2017. I hasn't made it into a release yet, but I assume it will at some point.

**Comment by cheeseplus** [comment]
Closed as fixed as there have been at least 2 releases since that time.
